elasticsearch/helpers/vectorstore/_async/vectorstore.py

```diff
@@ -0,0 +1,384 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
+#  license agreements. See the NOTICE file distributed with
+#  this work for additional information regarding copyright
+#  ownership. Elasticsearch B.V. licenses this file to you under
+#  the Apache License, Version 2.0 (the "License"); you may
+#  not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+import logging
+import uuid
+from typing import Any, Callable, Dict, List, Optional
+
+from elasticsearch import AsyncElasticsearch
+from elasticsearch._version import __versionstr__ as lib_version
+from elasticsearch.helpers import BulkIndexError, async_bulk
+from elasticsearch.helpers.vectorstore._async.embedding_service import (
+    AsyncEmbeddingService,
+)
+from elasticsearch.helpers.vectorstore._async.strategies import AsyncRetrievalStrategy
+from elasticsearch.helpers.vectorstore._utils import maximal_marginal_relevance
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncVectorStore:
+    """VectorStore is a higher-level abstraction of indexing and search.
+    Users can pick from available retrieval strategies.
+
+    Documents are flat text documents. Depending on the strategy, vector embeddings are
+    - created by the user beforehand
+    - created by this class in Python
+    - created in-stack by inference pipelines.
+    """
+
+    def __init__(
+        self,
+        es_client: AsyncElasticsearch,
+        index_name: str,
+        retrieval_strategy: AsyncRetrievalStrategy,
+        embedding_service: Optional[AsyncEmbeddingService] = None,
+        num_dimensions: Optional[int] = None,
+        text_field: str = "text_field",
+        vector_field: str = "vector_field",
+        metadata_mappings: Optional[Dict[str, Any]] = None,
+        user_agent: str = f"es-py-vs/{lib_version}",
```

Comment:
Are we planning for this user agent to only represent standalone use of this library, and for LangChain/LlamaIndex/etc. integrations to continue overwriting the agent with their own?

Reply:
Yes exactly. But we also want to track how many people do use this class directly.